### PR TITLE
Parent based sampling tests

### PR DESCRIPTION
**Pull request by Geal**
This extracts the tests for opentelemetry's parent based sampling that were introduced in #3109

**Checklist**

Complete the checklist (and note appropriate exceptions) before a final PR is raised.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [x] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]. It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]. Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]. Tick whichever testing boxes are applicable. If you are adding Manual Tests:
    - please document the manual testing (extensively) in the Exceptions.
    - please raise a separate issue to automate the test and label it (or ask for it to be labeled) as `manual test`

### CODE REVIEW

**Review by BrynCooke — CHANGES_REQUESTED** [review]
We need to add a tests for sampling. This should have a 50% rate and verify in jaeger that no half traces exist. 
Separate tests are needed for local and remote roots.

**Comment by Geal** [comment]
> We need to add a tests for sampling
these tests are useful in their current state and should be merged. Adding a sampling rate is likely to be more work to get it right (high potential of flaky test), that can be done as a follow up

**Comment by BrynCooke** [comment]
I think it's worth the effort if we want to try the prefiltering thing again. I would be OK with merging as long as any prefiltering follow up does include these tests.

Here's the reason:

If no layer registers interest in a span event, tracing registry will globally disable the span. This means that the span is not created at all rather than it just being prefiltered for a particular layer.
Subsequent child spans will then look like root spans and risk making the sampling decision again.

The work for this test needs to be done before any solution can be merged so unless there is a good reason to defer it I'd try and add it here for the same reason that these tests have been split out from the original PR that these came from.

**Comment by Geal** [comment]
> If your proposal is that we merge this stuff, and then make another PR for the tests and then a follow up PR for any prefiltering that that is OK.

Right now I have a good idea of how I will fix prefiltering, but no idea how to make a non flaky test around random based sampling. So I will work on the prefiltering fix, but we won't merge it until we have a test for the sampling

**Review by BrynCooke — APPROVED** [review]

**Comment by Geal** [comment]
it looks like the tests local_root, remote_root and reload never passed , looking into it

**Comment by Geal** [comment]
it appears the tests are not independent. If I run them individually, they pass, but as a whole they will fail

**Comment by BrynCooke** [comment]
I figured out what the issue was as part of #3364
Once this is merged this PR will need to be rebased, but then things will work.

**Comment by Geal** [comment]
ok so after a rebase the tests are now failing even individually

**Comment by BrynCooke** [comment]
It required more than a rebased in the end. Tests should be passing now.

**Comment by Geal** [comment]
great, thanks!

File: apollo-router/tests/jaeger_test.rs
Comment (o0Ignition0o) [inline apollo-router/tests/jaeger_test.rs:270]:
```suggestion
    tracing::debug!("{}", serde_json::to_string_pretty(&trace)?);
```

**Review by o0Ignition0o — APPROVED** [review]

**Comment by o0Ignition0o** [comment]
I m not sure if any action is needed wrt [this comment](https://github.com/apollographql/router/pull/3136/files#r1282179670) and the need of percent based sampling. otherwise lgtm.

**Pull request merged**
